### ISSUE TITLE
Reverse proxy mode

### ISSUE THREAD

**Issue by Manouchehri** [opticdev/optic#1857]
**Describe the enhancement**
In some situations, like serverless functions, setting a proxy isn't trivial; setting a different server URL would be trivial though!

**Additional information**
https://docs.mitmproxy.org/stable/concepts-modes/#reverse-proxy

Example usage:
```
dave@mbp corellium-openapi % optic capture openapi.json https://moda.enterprise.corellium.com/api --reverse-proxy
» Reverse proxy running on https://localhost:8000.
 help  Press [ Enter ] to finish capturing requests
⠋ 2 requests captured
```

And the usage would look like:
```
curl -kv "https://localhost:8000/api/example"
```

**Comment by niclim** [comment]
Hi, thanks for using Optic and the issue. We just released `0.42.12` which includes the `--reverse-proxy` command for capture which should work as you described.

_(this was actually under `--no-system-proxy` but we renamed it to `--reverse-proxy` as its easier to understand / more discoverable)_

**Comment by Manouchehri** [comment]
Hmm, mind testing this for me? I can't get it to work.

```sh
optic capture --reverse-proxy openapi.json https://moda.enterprise.corellium.com
```

```
dave@mbp corellium-openapi % optic capture --reverse-proxy openapi.json https://moda.enterprise.corellium.com
» Proxy running on https://localhost:8000. System proxy updated
 help  Press [ Enter ] to finish capturing requests

Not seeing any traffic captured? Make sure your HTTP Client is using the proxy: https://www.useoptic.com/docs/oas-reference/client-guides
⠼ 2 requests captured
```

```
dave@mbp ~ % curl -kv "https://localhost:8000/api/v1/ready"
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
* [CONN-0-0][CF-SSL] (304) (OUT), TLS handshake, Client hello (1):
* [CONN-0-0][CF-SSL] (304) (IN), TLS handshake, Server hello (2):
* [CONN-0-0][CF-SSL] (304) (IN), TLS handshake, Unknown (8):
* [CONN-0-0][CF-SSL] (304) (IN), TLS handshake, Certificate (11):
* [CONN-0-0][CF-SSL] (304) (IN), TLS handshake, CERT verify (15):
* [CONN-0-0][CF-SSL] (304) (IN), TLS handshake, Finished (20):
* [CONN-0-0][CF-SSL] (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-AES256-GCM-SHA384
* ALPN: server accepted http/1.1
* Server certificate:
*  subject: CN=localhost; C=XX; L=Unknown; O=Mockttp Cert - DO NOT TRUST
*  start date: Apr 19 21:26:36 2023 GMT
*  expire date: Apr 20 21:26:36 2024 GMT
*  issuer: CN=Optic CLI CA (locally generated 2023-04-20T14:57:04.005Z); C=US; ST=NY; L=New York City; O=Optic Labs Corporation; OU=https://useoptic.com
*  SSL certificate verify ok.
> GET /api/v1/ready HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.87.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< Server: awselb/2.0
< Date: Thu, 20 Apr 2023 21:27:29 GMT
< Content-Type: text/html
< Content-Length: 134
< Connection: keep-alive
< Location: https://localhost:443/api/v1/ready
<
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
</body>
</html>
```

Actual endpoint response:

```
dave@mbp ~ % curl -kv https://moda.enterprise.corellium.com/api/v1/ready
*   Trying 3.135.131.114:443...
* Connected to moda.enterprise.corellium.com (3.135.131.114) port 443 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
* [CONN-0-0][CF-SSL] (304) (OUT), TLS handshake, Client hello (1):
* [CONN-0-0][CF-SSL] (304) (IN), TLS handshake, Server hello (2):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS handshake, Certificate (11):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS handshake, Server finished (14):
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* [CONN-0-0][CF-SSL] TLSv1.2 (OUT), TLS handshake, Finished (20):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* [CONN-0-0][CF-SSL] TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=enterprise.corellium.com
*  start date: Feb  9 00:00:00 2023 GMT
*  expire date: Sep 20 23:59:59 2023 GMT
*  issuer: C=US; O=Amazon; CN=Amazon RSA 2048 M01
*  SSL certificate verify ok.
* Using HTTP2, server supports multiplexing
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* h2h3 [:method: GET]
* h2h3 [:path: /api/v1/ready]
* h2h3 [:scheme: https]
* h2h3 [:authority: moda.enterprise.corellium.com]
* h2h3 [user-agent: curl/7.87.0]
* h2h3 [accept: */*]
* Using Stream ID: 1 (easy handle 0x12c812800)
> GET /api/v1/ready HTTP/2
> Host: moda.enterprise.corellium.com
> user-agent: curl/7.87.0
> accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
< HTTP/2 204
< date: Thu, 20 Apr 2023 21:28:32 GMT
< strict-transport-security: max-age=63072000; includeSubDomains
< content-security-policy: frame-ancestors 'self'
< access-control-allow-origin: *
< server: nginx/1.14.0 (Ubuntu)
< x-powered-by: Express
< etag: W/"2-vyGp6PvFo4RvsFtPoIWeCReyIC8"
<
* Connection #0 to host moda.enterprise.corellium.com left intact
* ```

**Comment by Manouchehri** [comment]
See https://github.com/opticdev/optic/commit/cf4788251889db45f4dc01299e8eb76e81e16e07#r109900174.

**Comment by niclim** [comment]
Ugh ok, sorry that was a silly mistake! I'm building a fix now

If you try it with the conditional inverted do you still get 301 redirected via the reverse proxy? We might not be forwarding headers or something else (need to dig into this a little)

**Comment by niclim** [comment]
Ok, so just an update here - I can recreate this issue and have fixed one part of this (https://github.com/opticdev/optic/pull/1866) - I need to do some more investigating as the forwarded request is rejected by certain servers.

**Comment by niclim** [comment]
I just released `0.42.14`  which should fix issues with reverse-proxying to remote servers - there were issues with:
- https redirection (returned 301s)
- not rewriting the host header correctly

Let me know if these fixes work for you!
